### PR TITLE
[Enhancement] Adjust base version before publish version

### PR DESCRIPTION
**Pull request by TszKitLo40**
## Why I'm doing:
In batch publish, the number of txns may be increased each time. For example, the list of txns to be published is [11, 12], `base_version` is 1, `new_version` is 2. And BE published txn 1 successfully, the index version has increased to 2. But the next time when FE retries, the list of txns may be [11, 12, 13], `base_version` is 1, `new_version` is 3. The index version is larger than the `base_version` so BE will return failure which will make FE retry again.

## What I'm doing:
For the second time when FE retries the publish request, BE will adjust the base version according to index version and tablet metadata version and publish from the adjusted base version.
Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/storage/lake/transactions.cpp [hunk 1]
```diff
@@ -53,9 +71,35 @@ static void clear_remote_snapshot_async(TabletManager* tablet_mgr, int64_t table
...
+    if (metadata != nullptr) {
+        version = std::max(version, metadata->version());
+    }
+
+    auto index_version = tablet_mgr->update_mgr()->get_primary_index_data_version(tablet_id);
+    if (index_version > 0) {
+        auto res = tablet_mgr->get_tablet_metadata(tablet_id, index_version);
+        if (res.ok()) {
+            version = std::max(version, index_version);
+        }
```
Comment (sduzh) [inline be/src/storage/lake/transactions.cpp:86]: else?

File: be/src/storage/lake/transactions.cpp [hunk 2]
```diff
@@ -53,9 +71,35 @@ static void clear_remote_snapshot_async(TabletManager* tablet_mgr, int64_t table
...
 
+void adjust_base_version(int64_t tablet_id, TabletManager* tablet_mgr, int64_t* base_version) {
+    int64_t version = 0;
+    auto metadata = tablet_mgr->get_latest_cached_tablet_metadata(tablet_id);
+    if (metadata != nullptr) {
+        version = std::max(version, metadata->version());
+    }
+
+    auto index_version = tablet_mgr->update_mgr()->get_primary_index_data_version(tablet_id);
+    if (index_version > 0) {
```
Comment (sduzh) [inline be/src/storage/lake/transactions.cpp:82]:
```suggestion
    if (index_version > base_version) {
```

Comment (sduzh) [reply be/src/storage/lake/transactions.cpp:82]: Reduce unnecessary `get_tablet_metadata()` operations.

File: be/src/storage/lake/transactions.cpp [hunk 3]
```diff
@@ -53,9 +71,35 @@ static void clear_remote_snapshot_async(TabletManager* tablet_mgr, int64_t table
     files_to_delete->emplace_back(std::move(slog_path));
 }
 
+void adjust_base_version(int64_t tablet_id, TabletManager* tablet_mgr, int64_t* base_version) {
+    int64_t version = 0;
```
Comment (sduzh) [inline be/src/storage/lake/transactions.cpp:75]:
```suggestion
    int64_t version = *base_version;
```

File: be/src/storage/lake/transactions.cpp [hunk 4]
```diff
@@ -53,9 +71,37 @@ static void clear_remote_snapshot_async(TabletManager* tablet_mgr, int64_t table
...
+    auto metadata = tablet_mgr->get_latest_cached_tablet_metadata(tablet_id);
+    if (metadata != nullptr) {
+        version = std::max(version, metadata->version());
+    }
+
+    auto index_version = tablet_mgr->update_mgr()->get_primary_index_data_version(tablet_id);
+    if (index_version > version) {
+        auto res = tablet_mgr->get_tablet_metadata(tablet_id, index_version);
+        if (res.ok()) {
+            version = std::max(version, index_version);
```
Comment (sduzh) [inline be/src/storage/lake/transactions.cpp:85]:
```suggestion
            version = index_version;
```
You've already checked above that index_version > version

File: be/src/storage/lake/transactions.cpp [hunk 5]
```diff
@@ -53,9 +71,37 @@ static void clear_remote_snapshot_async(TabletManager* tablet_mgr, int64_t table
...
+        *base_version = version;
+        LOG(INFO) << "base version has been adjusted to " << *base_version;
+    }
+}
+
 StatusOr<TabletMetadataPtr> publish_version(TabletManager* tablet_mgr, int64_t tablet_id, int64_t base_version,
                                             int64_t new_version, std::span<const int64_t> txn_ids,
                                             int64_t commit_time) {
+    if (!add_tablet(tablet_id)) {
+        return Status::ResourceBusy(fmt::format("Does not support concurrent publishing, tablet {}", tablet_id));
```
Comment (sduzh) [inline be/src/storage/lake/transactions.cpp:103]:
```suggestion
        return Status::ResourceBusy(fmt::format("The previous publish version task for tablet {} has not finished. You can ignore this error and the task will retry later.", tablet_id));
```

File: be/src/storage/lake/transactions.cpp [hunk 6]
```diff
@@ -109,7 +165,8 @@ StatusOr<TabletMetadataPtr> publish_version(TabletManager* tablet_mgr, int64_t t
     // 5. txn4 will be published in later publish task, but we can't judge what's the latest_version in BE and we can not reapply txn_log if
     // txn logs have been deleted.
     bool delete_txn_log = (txn_ids.size() == 1);
-    for (int i = 0; i < txn_ids.size(); i++) {
+    int txn_offset = ori_base_version < base_version ? base_version - ori_base_version : 0;
```
Comment (sduzh) [inline be/src/storage/lake/transactions.cpp:168]:
```suggestion
    int txn_offset = base_version - ori_base_version;
```

File: be/src/storage/lake/transactions.cpp [hunk 7]
```diff
@@ -109,7 +165,8 @@ StatusOr<TabletMetadataPtr> publish_version(TabletManager* tablet_mgr, int64_t t
...
+    for (size_t i = txn_offset; i < txn_ids.size(); i++) {
```
Comment (sduzh) [inline be/src/storage/lake/transactions.cpp:169]: Should also update the if condition at lline 175?

**Review by sduzh — DISMISSED** [review]

File: be/src/storage/lake/transactions.cpp (hunk 7 above)
Comment (TszKitLo40) [reply be/src/storage/lake/transactions.cpp:172]: @smartlxh PTAL

File: be/src/storage/lake/transactions.cpp [hunk 8]
```diff
@@ -78,6 +129,14 @@ StatusOr<TabletMetadataPtr> publish_version(TabletManager* tablet_mgr, int64_t t
         return error;
     };
 
+    int64_t ori_base_version = base_version;
+    adjust_base_version(tablet_id, tablet_mgr, &base_version);
+    if (base_version > new_version) {
```
Comment (smartlxh) [inline be/src/storage/lake/transactions.cpp:134]: Should judge whether publish mode is batch by txns_size。If convert publish mode from batch to single and retry，the base version may greater than new_version

File: be/src/storage/lake/transactions.cpp (hunk 7 above)
Comment (smartlxh) [reply be/src/storage/lake/transactions.cpp:172]: The txn log should not be not found after `adjust_base_version` and an error should be reported directly

**Review by sduzh — DISMISSED** [review]

**Review by sduzh — DISMISSED** [review]

**Review by smartlxh — APPROVED** [review]

File: be/src/storage/lake/transactions.cpp [hunk 9]
```diff
@@ -53,9 +71,43 @@ static void clear_remote_snapshot_async(TabletManager* tablet_mgr, int64_t table
...
+int64_t cal_new_base_version(int64_t tablet_id, TabletManager* tablet_mgr, int64_t base_version, int64_t new_version) {
+    int64_t version = base_version;
+    auto metadata = tablet_mgr->get_latest_cached_tablet_metadata(tablet_id);
+    if (metadata != nullptr && metadata->version() <= new_version) {
+        version = std::max(version, metadata->version());
+    }
+
+    auto index_version = tablet_mgr->update_mgr()->get_primary_index_data_version(tablet_id);
+    if (index_version > new_version) {
+        tablet_mgr->update_mgr()->try_remove_primary_index_cache(tablet_id);
```
Comment (luohaha) [inline be/src/storage/lake/transactions.cpp:83]: If `try_remove_primary_index_cache` fail because `ref` no equal 1, does it matters?

Comment (luohaha) [reply be/src/storage/lake/transactions.cpp:83]: And why no check whether the index version exists in remote storage. If not, clear and rebuild the index

File: be/src/storage/lake/update_manager.cpp [hunk 10]
```diff
@@ -129,6 +129,17 @@ void UpdateManager::remove_primary_index_cache(IndexEntry* index_entry) {
     }
 }
 
+void UpdateManager::remove_primary_index(int64_t tablet_id) {
```
Comment (luohaha) [inline be/src/storage/lake/update_manager.cpp:132]: How about `remove_unload_primary_index` ?

**Review by luohaha — DISMISSED** [review]

**Review by luohaha — APPROVED** [review]

**Review by sduzh — APPROVED** [review]

**Pull request merged**